src/node_report.cc

```diff
@@ -0,0 +1,1024 @@
+                            Environment* env,
+                            const char* message,
+                            const char* location,
+                            const std::string& filename,
+                            std::ostream& out,
+                            MaybeLocal<Value> error,
+                            TIME_TYPE* tm_struct) {
+  std::ostringstream buf;
+  int indent = 0;
+  PID_TYPE pid = uv_os_getpid();
+
+  // Save formatting for output stream.
+  std::ios old_state(nullptr);
+  old_state.copyfmt(out);
+
+  // File stream opened OK, now start printing the report content:
+  // the title and header information (event, filename, timestamp and pid)
+  // PrintSectionHeader(out, "Node Report");
+
+  json_start(out, indent);
+  json_indent(&indent);
+  json_objectstart(out, indent, "header");
+  json_indent(&indent);
+
+  json_keyvaluenormal(out, indent, "event", message);
+  json_keyvaluenormal(out, indent, "location", location);
+  if (!filename.empty())
+    json_keyvaluenormal(out, indent, "filename", filename);
+  else
+    json_keyvaluenormal(out, indent, "filename", std::string("''"));
+
+  // Print dump event and module load date/time stamps
+  char timebuf[64];
+#ifdef _WIN32
+  snprintf(timebuf, sizeof(timebuf), "%4d/%02d/%02d %02d:%02d:%02d",
+          tm_struct->wYear, tm_struct->wMonth, tm_struct->wDay,
+          tm_struct->wHour, tm_struct->wMinute, tm_struct->wSecond);
+  json_keyvaluenormal(out, indent, "dump_event_time", timebuf);
+#else  // UNIX, OSX
+  snprintf(timebuf, sizeof(timebuf), "%4d/%02d/%02d %02d:%02d:%02d",
+          tm_struct->tm_year+1900, tm_struct->tm_mon+1, tm_struct->tm_mday,
+          tm_struct->tm_hour, tm_struct->tm_min, tm_struct->tm_sec);
+  json_keyvaluenormal(out, indent, "dump_event_time", timebuf);
+#endif
+  // Print native process ID
+  buf << pid;
+  json_keyvaluenormal(out, indent, "process_id", buf.str());
+  buf.flush();
+
+  // Print out the command line.
+  if (node::per_process_opts->cmdline != "") {
+    json_keyvaluenormal(
+        out, indent, "command_line", node::per_process_opts->cmdline);
+  }
+
+  // Print Node.js and OS version information
+  PrintVersionInformation(out, &indent);
+  json_deindent(&indent);
+  json_objectendnormal(out, indent);
+
+  // Print summary JavaScript stack backtrace
+  PrintJavaScriptStack(out, isolate, error, location, &indent);
+
+  // Print native stack backtrace
+  PrintNativeStack(out, &indent);
+
+  // Print V8 Heap and Garbage Collector information
+  PrintGCStatistics(out, isolate, &indent);
+
+  // Print OS and current thread resource usage
+#ifndef _WIN32
+  PrintResourceUsage(out, &indent);
+#endif
+
+  write_data data;
+  data.out = &out;
+  data.indent = &indent;
+  json_arraystart(out, indent, "libuv");
+  json_indent(&indent);
+  if (env != nullptr)
+    uv_walk(env->event_loop(), WalkHandle, static_cast<void*>(&data));
+  else
+    uv_walk(uv_default_loop(), WalkHandle, static_cast<void*>(&data));
```

Comment:
Oh, just noticing now… this would still not necessarily be thread-safe if `env != nullptr`… I don’t know if there’s a way to check, but you probably want to unly run this if you’re on the main thread?

Reply:
sure - I was checking in the code to see if there is a primitive that help me to figure out which thread  I am in, but could not find one. My thinking is to introduce a `thread_local` variable, initialize with the `pthread_self` value in the early phase where we know only main thread exists, and then check the current thread id against that at this call site.

Should we worry about this when we go for full worker support, and document lack of support to workers until such times?